com.unity.multiplayer.mlapi/Runtime/Core/NetworkObject.cs

```diff
@@ -41,7 +41,9 @@ private void OnValidate()
         /// <summary>
         /// Gets the NetworkManager that owns this NetworkObject instance
         /// </summary>
-        public NetworkManager NetworkManager => NetworkManager.Singleton;
+        public NetworkManager NetworkManager => NetworkManagerTestOverride != null ? NetworkManagerTestOverride : NetworkManager.Singleton;
+
+        internal NetworkManager NetworkManagerTestOverride;
```

Comment:
we should have `#if UNITY_EDITOR || DEVELOPMENT_BUILD` surrounding this.
(we wouldn't have `NetworkManagerTestOverride` in non-Editor, non-Development builds)

Reply:
Not too sure I agree tbh. This is internal behaviour which we might want exposed eventually. I generally prefer not to conditional compile things if we don't have to. 

Perhaps renaming it and removing the word "Test" from the field name?

Reply:
we're not there yet.
let's keep this more focused on testing for now.

Reply:
We're not there yet, but we want to move there as quickly as possible. Part of this epic is to prepare for that future when we are there. IFDefing is going against of the goal outlined in the brief.

Reply:
we're not there yet, today.
we're adding a very small overhead and extra member into release builds but in reality, it doesn't actually have to be there.

when the use-case and implementations in other places started to become more generic, then we can roll it out to other places as well.
```suggestion
#if UNITY_EDITOR || DEVELOPMENT_BUILD
        public NetworkManager NetworkManager => NetworkManagerTestOverride != null ? NetworkManagerTestOverride : NetworkManager.Singleton;

        internal NetworkManager NetworkManagerTestOverride;
#else
        public NetworkManager NetworkManager => NetworkManager.Singleton;
#endif
```

Reply:
The epic is not just about testing. It's also about preparing for full singleton-less support. This is why it should always be there.

Quote from the proposal:
> This partial step is also a clear path forward to remove Singleton dependency fully from the library and allow users to also run multiple NetworkManagers at once.

Reply:
are we going to keep this logic when we're living in a _full singleton-less_ world?

- check if `NetworkManagerTestOverride` member is `null`
  - if `null`, return `NetworkManager.Singleton`
  - if not, return `NetworkManagerTestOverride`

in other terms, which pieces of code you're seeing in this current code block will stay there in the future?

Reply:
Probably will stay yes, that's why I suggested the removal of the Test in the name. The full singleton less world is not defined yet. But I can imagine one solution where the Singleton property goes and we just use that field which is in turn set by some system that controls what belongs to what.

The proposal wants to minimize breakages and prepare for a singleton less future. 

This accomplishes both.

Reply:
so would you agree that we'll most likely have something like this in the end:

```cs
public NetworkManager NetworkManager;
```

instead of

```cs
public NetworkManager NetworkManager;
internal NetworkManager NetworkManagerOverride;
```

Reply:
Yes that's how I envision it, but probably readonly property so the user cannot set it.

Reply:
well, in that case, `NetworkManagerOverride` serves a temporary purpose and that is my point.
I'd like to keep the name as it is: `NetworkManagerTestOverride` (because that makes it more specific).
I only want to wrap them under `#if UNITY_EDITOR || DEVELOPMENT_BUILD` macros because they'll be there for a while for test/dev purposes and eventually removed.
there is no reason to roll them over to release build runtime, at all.

Reply:
It does not serve a temporary purpose. As I said, later on, we can use that field to set where something should belong. So it is permanent, even though we might refactor the names.

We just can't have one field now called NetworkManager because in that case we had to check if its null, then fall back to singleton in all the callsites.

But say if .Singleton was removed, we could refactor it to be cleaner, but the field should stay.

Thus, it should be a part of the library, tests and more.

It's not a temporary thing, it's a permanent change.

Reply:
ultimately, after `.Singleton` removal, we'll probably have this:

```cs
public NetworkManager NetworkManager { get; internal set; }
// no NetworkManagerTestOverride or NetworkManagerOverride
```

today, we're adding an extra internal overriding member and falling back to `.Singleton` if we didn't override.
this specific override functionality is temporary and will go away — or in fact, we're only going to use this approach for testing/development for quite a while.
which brings me back to the original point, let's have it in the Editor and Development builds but not roll it over to release build runtimes, it has no other use today and eventually will be gone entirely in the future.

Reply:
We will not remove the override in the future. We are adding the override in this commit, and later on we will rename it. But the functionality of override will stay (internal set part). This is the commit that adds that functionality to the core library, not a future one. It is the task of this epic to do that.

Reply:
we'll use `NetworkManagerOverride` member at runtime in release builds, is that what you're saying?

Reply:
Yeah. That's part of this PR, to prepare everything to be able to run singleton less. We just aren't exposing the API for it.

Reply:
I'm looking at the internal gdoc proposal, where did you outline upcoming changes you're referring to? are they somewhere else?

Reply:
> I'm looking at the internal gdoc proposal, where did you outline upcoming changes you're referring to? are they somewhere else?

This part:
> This partial step is also a clear path forward to remove Singleton dependency fully from the library and allow users to also run multiple NetworkManagers at once.

is what this is doing. Preparing the library for that stage.

Reply:
yeah, I see. let's have a chat next week about this.

Reply:
I've been bitten by #ifdef blocked code enough times to err on the side of not #ifdef gating things when not necessary.

And I would wonder, can we count on DEVELOPMENT_BUILD being set when we're doing testing?  What if we're testing a fully production build?

But I would stipulate we decorate this variable however we name it with a comment, something to the effect of "in normal use, NetworkObject's all share the one-and-only-one networking manager.  For testing and some anticipated future uses, we allow this networking manager to be overridden so that objects can belong to multiple different NetworkManagers"

Reply:
I'm not sure why this conversation marked as resolved. I'm still not entirely happy about rolling something out over runtime without proper planning, especially if something was targeting testing/debugging use-cases only somehow started to effect runtime code.
please organize an internal meeting and let's try to align on our future plans together.